### ISSUE TITLE
🚀 Feature: SSE (Server Sent Events) support

### ISSUE THREAD

**Issue by JoelBonetR** [novuhq/novu#1137]
### 🔖 Feature description

Support for SSE

### 🎤 Why is this feature needed ?

SSE is best used when it's not necessary to send data from client to server. In status updates and push notification applications, the data flow -usually- follows a single path that is from the server to the client only.

This is what SSE is designed for, so WebSocket would be overkill as you are opening UDP bi-directional channels at each request/user which will consume server resources with all probability.

### ✌️ How do you aim to achieve this?

Replicating websocket implementation with SSE should be relatively easy. See [Server-Sent Events](https://docs.nestjs.com/techniques/server-sent-events) in Nest JS documentation.

### 🔄️ Additional Information

Websockets and SSE (Server Sent Events) are both capable of pushing data **to** browsers, however they are not competing technologies.

Websockets connections can both send data to the browser and receive data from the browser. A good example of an application that could use websockets is a chat application.

**SSE connections can only push data to the browser**. Notifications, Online stock quotes, or RRSS updating the timeline or feed are good examples of an application that could benefit from SSE.

In practice since everything that can be done with SSE can also be done with Websockets, however, it can be overkill for some types of application, and the server will consume less resources when using SSE.

SSE applies a _publish-subscribe pattern_ or _request-respond pattern_, despite the response is a stream.

**Browser support:**

[Server-sent events. Eventsource](https://caniuse.com/eventsource) ~97.97%

Furthermore SSE can be polyfilled into older browsers that do not support it natively using just JavaScript. Some implementations of SSE polyfills can be found on the [Moderniz github page](https://github.com/Modernizr/Modernizr/wiki/HTML5-Cross-Browser-Polyfills).

**Related Libs:**
- [EventSource](https://github.com/Yaffle/EventSource/blob/master/src/eventsource.js) by Yaffle.
- [EventSource](https://github.com/remy/polyfills/blob/master/EventSource.js) by Remi Sharp.
- [Nest JS - SSE](https://github.com/serkyron/nestjs-sse) by serkyron (Note that it's not using HTTP2).

**Limitations:**

- **Maximum open connections limit.** SSE over HTTP/1.1 suffers from a limitation to the maximum number of open connections, which can be specially painful when opening various tabs as the limit is per browser and set to a very low number (6). The issue has been marked as "Won't fix" in [Chrome](https://bugs.chromium.org/p/chromium/issues/detail?id=275955) and [Firefox](https://bugzilla.mozilla.org/show_bug.cgi?id=906896). This limit is per browser + domain, so that means that you can open 6 SSE connections across all of the tabs to www.mywebsite1.com and another 6 SSE connections to www.mywebsite2.com.
- When using **HTTP/2** (recommended), the maximum number of simultaneous HTTP streams is negotiated between the server and the client (defaults to 100).
- **No binary support.** While WS can transmit both binary data and UTF-8, SSE is limited to UTF-8.

**Extended information sources:**
- [EventSource basics](https://web.dev/eventsource-basics/)
- [EventSource MDN](https://developer.mozilla.org/en-US/docs/Web/API/EventSource)
- [WHATWG Specification](https://html.spec.whatwg.org/multipage/server-sent-events.html#the-eventsource-interface)

**TL;DR**
In Server-sent Events, the communication is carried out from server to browser only, the browser/client cannot send any data to the server through the same connection. This kind of communication is mainly used when the need is only to show the updated data, then the server sends the message whenever the data gets updated. It is a one-way communication from the Server to the client attached to a given event.

### 👀 Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Contributing Guidelines](https://github.com/novuhq/novu/blob/main/CONTRIBUTING.md)

**Comment by scopsy** [comment]
Thank you @JoelBonetR for the detailed writeup here, will take a deep look soon and will chat with the team about it aswell. <3

**Comment by JoelBonetR** [comment]
> Thank you @JoelBonetR for the detailed writeup here, will take a deep look soon and will chat with the team about it aswell. <3

Thank you @scopsy , I added this "minimal" (to say something 😅) [Pull Request as initial commit](https://github.com/novuhq/novu/pull/1140) on this topic if you want to extend the thingy from here. I tried to build something bigger but I noticed a lack of knowledge about how is that supposed to work inside the Novu implementation.

Ping me on Discord if anyone need support on the topic, I'll try my best.

**Comment by scopsy** [comment]
Amazing, I will be taking a look at the PR soon. Been a really busy week here 😄

**Comment by JoelBonetR** [comment]
Thank you! :)

**Comment by kymppi** [comment]
Any progress on this?

**Comment by scopsy** [comment]
I think that for now, we will leave it out since we are focusing on other parts of the system. I will try to get back to it in a few weeks/months to take a closer look at this proposal when we have a bit more bandwidth.
